transport/transport-runtime/src/test/java/com/google/android/datatransport/runtime/scheduling/persistence/SchemaManagerTest.java

```diff
@@ -71,4 +71,103 @@ public void persist_correctlyRoundTrips() {
     assertThat(newEvent.getEvent()).isEqualTo(EVENT1);
     assertThat(events).containsExactly(newEvent);
   }
+
+  @Test
+  public void upgradingV1ToV2_emptyDatabase_allowsPersistsAfterUpgrade() {
+    int oldVersion = 1;
+    int newVersion = 2;
+    SchemaManager schemaManager = new SchemaManager(RuntimeEnvironment.application, oldVersion);
+
+    SQLiteEventStore store = new SQLiteEventStore(clock, new UptimeClock(), CONFIG, schemaManager);
+
+    schemaManager.onUpgrade(schemaManager.getWritableDatabase(), oldVersion, newVersion);
+    PersistedEvent newEvent1 = store.persist(CONTEXT1, EVENT1);
+
+    assertThat(store.loadBatch(CONTEXT1)).containsExactly(newEvent1);
+  }
+
+  @Test
+  public void upgradingV1ToV2_nonEmptyDB_isLossless() {
+    int oldVersion = 1;
+    int newVersion = 2;
+    SchemaManager schemaManager = new SchemaManager(RuntimeEnvironment.application, oldVersion);
+    SQLiteEventStore store = new SQLiteEventStore(clock, new UptimeClock(), CONFIG, schemaManager);
+    // We simulate operations as done by an older SQLLiteEventStore at V1
+    // We cannot simulate older operations with a newer client
+    PersistedEvent event1 = simulatedPersistOnV1Database(schemaManager, CONTEXT1, EVENT1);
+
+    // Upgrade to V2
+    schemaManager.onUpgrade(schemaManager.getWritableDatabase(), oldVersion, newVersion);
+
+    assertThat(store.loadBatch(CONTEXT1)).containsExactly(event1);
+  }
+
+  @Test
+  public void downgradeV2ToV1_withEmptyDB_allowsPersistanceAfterMigration() {
+    int fromVersion = 2;
+    SchemaManager schemaManager = new SchemaManager(RuntimeEnvironment.application, fromVersion);
+
+    SQLiteEventStore store = new SQLiteEventStore(clock, new UptimeClock(), CONFIG, schemaManager);
+    // We simulate operations as done by an older SQLLiteEventStore at V1
+    // We cannot simulate older operations with a newer client
+    simulatedPersistOnV1Database(schemaManager, CONTEXT1, EVENT1);
+
+    schemaManager.onDowngrade(schemaManager.getWritableDatabase(), fromVersion, -1);
+    PersistedEvent event2 = store.persist(CONTEXT2, EVENT2);
+
+    assertThat(store.loadBatch(CONTEXT2)).containsExactly(event2);
+  }
+
+  @Test
+  public void downgradeV2ToV1_withNonEmptyDB_isLossy() {
+    int fromVersion = 2;
+    int toVersion = fromVersion - 1;
+    SchemaManager schemaManager = new SchemaManager(RuntimeEnvironment.application, fromVersion);
+    SQLiteEventStore store = new SQLiteEventStore(clock, new UptimeClock(), CONFIG, schemaManager);
+    PersistedEvent event1 = store.persist(CONTEXT1, EVENT1);
+
+    schemaManager.onDowngrade(schemaManager.getWritableDatabase(), toVersion, fromVersion);
+
+    assertThat(store.loadBatch(CONTEXT1)).doesNotContain(event1);
+  }
+
+  @Test(expected = RuntimeException.class)
```

Comment:
It's [not recommended](go/java-testing/exceptions) to use `@Test(expected)` as there is no way to tell at which point the exception was actually thrown, use `Assert.assertThrows` from junit.

Reply:
Took this approach since I saw other SDKs vendoring it in and I was lazy